Avoid raising an exception if returned latency is -1

ahothan opened the pull request:
In case of heavy packet drops, it is possible that no latency packets return, which 
will cause the returned latency value to be -1.
In this case, it is not desirable to have an exception in the python library and instead returns
-1 latency.

hhaim commented:
Hi Alec, 
I've reverted the commit, 

1. You fixed it in the wrong line. It should be in the latency test. 
2. The fix in not clear. Could you dump the rel_val and current_val to see why the value is -1?

```
 if 'err_cntrs' in pg_id_val and 'err_cntrs' in self.ref['latency'][pg_id]:
                    for key, val in pg_id_val['err_cntrs'].items():
                        ref_val = self.ref['latency'][pg_id]['err_cntrs'].get(key)
                        if ref_val is not None:
                            rel_val = val - ref_val
                            assert rel_val >= 0, 'Negative latency error value: %s (%s %s)' % (rel_val, pg_id, key)
                            pg_id_val['err_cntrs'][key] = rel_val
```

ahothan commented:
Hi Hanoh,

Sorry for the mistake…
Here is another repro with some additional traces on the received latency dict.

It looks like for pg_id 257, the dropped count went down from 20483 to 20481 between the ref and the received dict.
Actually a few more pg_ids also have their drop count decreased.

Also in some pg_ids, the last_max or total_max is not set, which looks like a missing assignment bug.


2019-08-25 09:12:49,870 INFO Running traffic generator
2019-08-25 09:12:59,960 INFO TX: 185263440; RX: 3504; Est. Dropped: 185259936; Est. Drop rate: 99.9981%
2019-08-25 09:13:10,003 INFO TX: 372036448; RX: 6943; Est. Dropped: 372029505; Est. Drop rate: 99.9981%
2019-08-25 09:13:20,046 INFO TX: 558845208; RX: 10398; Est. Dropped: 558834810; Est. Drop rate: 99.9981%
2019-08-25 09:13:30,091 INFO TX: 745647508; RX: 14920; Est. Dropped: 745632588; Est. Drop rate: 99.9980%
2019-08-25 09:13:40,136 INFO TX: 932467556; RX: 18416; Est. Dropped: 932449140; Est. Drop rate: 99.9980%
2019-08-25 09:13:49,979 INFO TX: 1115257118; RX: 22741; Est. Dropped: 1115234377; Est. Drop rate: 99.9980%
2019-08-25 09:13:52,015 INFO Average drop rate: 99.997781
2019-08-25 09:13:52,015 INFO Range search [0.0 .. 25.0] targets: {'ndr': 0.001, 'pdr': 0.1}
2019-08-25 09:13:52,015 INFO Modified traffic stream for port 0, new rate=12.5%.
2019-08-25 09:13:52,016 INFO Modified traffic stream for port 1, new rate=12.5%.
2019-08-25 09:13:52,016 INFO Running traffic generator

--------self.ref latency:
{256: {'err_cntrs': {'dropped': 73460,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 3,
                     'seq_too_low': 0},
       'latency': {'average': 286555.0,
                   'hdrh': u'HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEisL2Kaasc00ImJgCC8Qbe',
                   'histogram': {100000: 1, 400000: 2},
                   'jitter': 62897,
                   'last_max': 407681,
                   'total_max': 407681,
                   'total_min': 100000}},
257: {'err_cntrs': {'dropped': 20483,        <<<<<<<<< decreased to 20481
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 54160.0,
                   'hdrh': u'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASKwvYoJAFr8BV4=',
                   'histogram': {100000: 1},
                   'jitter': 6770,
                   'last_max': 0,         <<<<<< last_max not set
                   'total_max': 156434,
                   'total_min': 100000}},
258: {'err_cntrs': {'dropped': 27004,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 3,
                     'seq_too_low': 0},
       'latency': {'average': 314954.125,
                   'hdrh': u'HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEisL2K6TIX03opJgCDmwcO',
                   'histogram': {100000: 2, 500000: 1},
                   'jitter': 31196,
                   'last_max': 516308,
                   'total_max': 155348,
                   'total_min': 100000}},
259: {'err_cntrs': {'dropped': 101844,        <<<<<< decreased to 27470
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 3,
                     'seq_too_low': 0},
       'latency': {'average': 164739.0,
                   'hdrh': u'HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEi0N7KyNTMwuTLBAB+GgYY',
                   'histogram': {100000: 1, 200000: 2},
                   'jitter': 13988,
                   'last_max': 172618,
                   'total_max': 154324,
                   'total_min': 100000}},
384: {'err_cntrs': {'dropped': 94023,         <<<<< decreased to 0
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 82044.5,
                   'hdrh': u'HISTFAAAACN4nJNpmSzMwMDAwgABzFCaEUzOmNZg/wEi0NjCyAQAX8UFNA==',
                   'histogram': {100000: 1},
                   'jitter': 10255,
                   'last_max': 0,
                   'total_max': 0,           <<<<< last_max and total_max not set
                   'total_min': 100000}},
385: {'err_cntrs': {'dropped': 34780,       <<<< decreased to 0
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 44811.0,
                   'hdrh': u'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASLwvpQJAFuaBZE=',
                   'histogram': {80000: 1},
                   'jitter': 5601,
                   'last_max': 0,
                   'total_max': 0,         <<<<<<< last_max and total_max not set
                   'total_min': 80000}},
386: {'err_cntrs': {'dropped': 37343,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 71582.5,
                   'hdrh': u'HISTFAAAACN4nJNpmSzMwMDAwgABzFCaEUzOmNZg/wEisLuRkQkAYKQFaw==',
                   'histogram': {100000: 1},
                   'jitter': 8947,
                   'last_max': 0,
                   'total_max': 0,          <<<<<<<< last_max and total_max not set
                   'total_min': 100000}},
387: {'err_cntrs': {'dropped': 114025,
                     'dup': 0,
                     'out_of_order': 1,
                     'seq_too_high': 2,
                     'seq_too_low': 1},
       'latency': {'average': 170685.75,
                   'hdrh': u'HISTFAAAACl4nJNpmSzMwMDAxQABzFCaEUzOmNZg/wEi0NzIyPSYlWmpGBMAh4gG4A==',
                   'histogram': {100000: 2, 400000: 1},
                   'jitter': 45641,
                   'last_max': 0,
                   'total_max': 0,           <<<<<<< last_max and total_max not set
                   'total_min': 100000}},
'global': {'bad_hdr': 0, 'old_flow': 0}}

--------received latency:
{256: {'err_cntrs': {'dropped': 73460,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 3,
                     'seq_too_low': 0},
       'latency': {'average': 286555.0,
                   'hdrh': u'HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEisL2Kaasc00ImJgCC8Qbe',
                   'histogram': {100000: 1, 400000: 2},
                   'jitter': 62897,
                   'last_max': 407681,
                   'total_max': 407681,
                   'total_min': 100000}},
257: {'err_cntrs': {'dropped': 20481,
                     'dup': 0,
                     'out_of_order': 2,
                     'seq_too_high': 0,
                     'seq_too_low': 2},
       'latency': {'average': 130555.0,
                   'hdrh': u'HISTFAAAACd4nJNpmSzMwMDAwQABzFCaEUzOmNZg/wEisL2KaS8HkyATAHqaBj0=',
                   'histogram': {100000: 2},
                   'jitter': 8774,
                   'last_max': 156434,
                   'total_max': 156434,
                   'total_min': 100000}},
258: {'err_cntrs': {'dropped': 98274,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 235151.0625,
                   'hdrh': u'HISTFAAAACt4nJNpmSzMwMDAwwABzFCaEUzOmNZg/wEisL2KaT8H00QmpvVSTACZnweQ',
                   'histogram': {100000: 1},
                   'jitter': 51806,
                   'last_max': 155348,
                   'total_max': 155348,
                   'total_min': 100000}},
259: {'err_cntrs': {'dropped': 27470,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 159531.5,
                   'hdrh': u'HISTFAAAACt4nJNpmSzMwMDAwwABzFCaEUzOmNZg/wEi8LKJkWk2E1MzC5MvEwCXygcZ',
                   'histogram': {100000: 1},
                   'jitter': 14257,
                   'last_max': 154324,
                   'total_max': 154324,
                   'total_min': 100000}},
384: {'err_cntrs': {'dropped': 0,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 0,
                     'seq_too_low': 0},
       'latency': {'average': 82044.5,
                   'hdrh': u'HISTFAAAACN4nJNpmSzMwMDAwgABzFCaEUzOmNZg/wEi0NjCyAQAX8UFNA==',
                   'histogram': {},
                   'jitter': 10255,
                   'last_max': 0,
                   'total_max': 0,
                   'total_min': 100000}},
385: {'err_cntrs': {'dropped': 0,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 0,
                     'seq_too_low': 0},
       'latency': {'average': 44811.0,
                   'hdrh': u'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASLwvpQJAFuaBZE=',
                   'histogram': {},
                   'jitter': 5601,
                   'last_max': 0,
                   'total_max': 0,
                   'total_min': 80000}},
386: {'err_cntrs': {'dropped': 37343,
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,
                     'seq_too_low': 0},
       'latency': {'average': 71582.5,
                   'hdrh': u'HISTFAAAACN4nJNpmSzMwMDAwgABzFCaEUzOmNZg/wEisLuRkQkAYKQFaw==',
                   'histogram': {100000: 1},
                   'jitter': 8947,
                   'last_max': 0,
                   'total_max': 0,
                   'total_min': 100000}},
387: {'err_cntrs': {'dropped': 114025,
                     'dup': 0,
                     'out_of_order': 1,
                     'seq_too_high': 2,
                     'seq_too_low': 1},
       'latency': {'average': 170685.75,
                   'hdrh': u'HISTFAAAACl4nJNpmSzMwMDAxQABzFCaEUzOmNZg/wEi0NzIyPSYlWmpGBMAh4gG4A==',
                   'histogram': {100000: 2, 400000: 1},
                   'jitter': 45641,
                   'last_max': 0,
                   'total_max': 0,
                   'total_min': 100000}},
'global': {'bad_hdr': 0, 'old_flow': 0}}
2019-08-25 09:14:02,137 INFO Clean up skipped.
2019-08-25 09:14:02,258 ERROR {'status': 'ERROR', 'error_message': 'Traceback (most recent call last):\n  File "/nfvbench/nfvbench/nfvbench.py", line 644, in main\n    raise Exception(result[\'error_message\'])\nException: Traceback (most recent call last):\n  File "/nfvbench/nfvbench/nfvbench.py", line 123, in run\n    "service_chain": self.chain_runner.run(),\n  File "/nfvbench/nfvbench/chain_runner.py", line 175, in run\n    results[self.chain_name] = {\'result\': self.__get_chain_result()}\n  File "/nfvbench/nfvbench/chain_runner.py", line 147, in __get_chain_result\n    self.config.traffic.bidirectional))\n  File "/nfvbench/nfvbench/chain_runner.py", line 123, in __get_result_per_frame_size\n    results = self.traffic_client.get_ndr_and_pdr()\n  File "/nfvbench/nfvbench/traffic_client.py", line 710, in get_ndr_and_pdr\n    self.__range_search(0.0, 200.0, targets, results)\n  File "/nfvbench/nfvbench/traffic_client.py", line 883, in __range_search\n    self.__range_search(left, middle, left_targets, results)\n  File "/nfvbench/nfvbench/traffic_client.py", line 883, in __range_search\n    self.__range_search(left, middle, left_targets, results)\n  File "/nfvbench/nfvbench/traffic_client.py", line 883, in __range_search\n    self.__range_search(left, middle, left_targets, results)\n  File "/nfvbench/nfvbench/traffic_client.py", line 846, in __range_search\n    stats, rates = self.__run_search_iteration(middle)\n  File "/nfvbench/nfvbench/traffic_client.py", line 901, in __run_search_iteration\n    for stats in self.run_traffic():\n  File "/nfvbench/nfvbench/traffic_client.py", line 941, in run_traffic\n    stats = self.runner.run()\n  File "/nfvbench/nfvbench/traffic_client.py", line 64, in run\n    return self.poll_stats()\n  File "/nfvbench/nfvbench/traffic_client.py", line 106, in poll_stats\n    return self.client.get_stats()\n  File "/nfvbench/nfvbench/traffic_client.py", line 739, in get_stats\n    stats = self.gen.get_stats()\n  File "/nfvbench/nfvbench/traffic_gen/trex_gen.py", line 778, in get_stats\n    stats = self.client.get_stats()\n  File "/usr/lib/python2.7/site-packages/trex/common/trex_api_annotators.py", line 51, in wrap2\n    ret = f(*args, **kwargs)\n  File "/usr/lib/python2.7/site-packages/trex/stl/trex_stl_client.py", line 1414, in get_stats\n    pgid_stats = self.get_pgid_stats()\n  File "/usr/lib/python2.7/site-packages/trex/common/trex_api_annotators.py", line 51, in wrap2\n    ret = f(*args, **kwargs)\n  File "/usr/lib/python2.7/site-packages/trex/stl/trex_stl_client.py", line 1591, in get_pgid_stats\n    return self.pgid_stats.get_stats(pgid_list)\n  File "/usr/lib/python2.7/site-packages/trex/stl/trex_stl_stats.py", line 301, in get_stats\n    assert rel_val >= 0, \'Negative latency error value: %s (%s %s) val=%d ref_val=%d\' % (rel_val, pg_id, key, val, ref_val)\nAssertionError: Negative latency error value: -2 (257 dropped) val=20481 ref_val=20483\n\n'}
Traceback (most recent call last):
  File "/nfvbench/nfvbench/nfvbench.py", line 123, in run
    "service_chain": self.chain_runner.run(),
  File "/nfvbench/nfvbench/chain_runner.py", line 175, in run
    results[self.chain_name] = {'result': self.__get_chain_result()}
  File "/nfvbench/nfvbench/chain_runner.py", line 147, in __get_chain_result
    self.config.traffic.bidirectional))
  File "/nfvbench/nfvbench/chain_runner.py", line 123, in __get_result_per_frame_size
    results = self.traffic_client.get_ndr_and_pdr()
  File "/nfvbench/nfvbench/traffic_client.py", line 710, in get_ndr_and_pdr
    self.__range_search(0.0, 200.0, targets, results)
  File "/nfvbench/nfvbench/traffic_client.py", line 883, in __range_search
    self.__range_search(left, middle, left_targets, results)
  File "/nfvbench/nfvbench/traffic_client.py", line 883, in __range_search
    self.__range_search(left, middle, left_targets, results)
  File "/nfvbench/nfvbench/traffic_client.py", line 883, in __range_search
    self.__range_search(left, middle, left_targets, results)
  File "/nfvbench/nfvbench/traffic_client.py", line 846, in __range_search
    stats, rates = self.__run_search_iteration(middle)
  File "/nfvbench/nfvbench/traffic_client.py", line 901, in __run_search_iteration
    for stats in self.run_traffic():
  File "/nfvbench/nfvbench/traffic_client.py", line 941, in run_traffic
    stats = self.runner.run()
  File "/nfvbench/nfvbench/traffic_client.py", line 64, in run
    return self.poll_stats()
  File "/nfvbench/nfvbench/traffic_client.py", line 106, in poll_stats
    return self.client.get_stats()
  File "/nfvbench/nfvbench/traffic_client.py", line 739, in get_stats
    stats = self.gen.get_stats()
  File "/nfvbench/nfvbench/traffic_gen/trex_gen.py", line 778, in get_stats
    stats = self.client.get_stats()
  File "/usr/lib/python2.7/site-packages/trex/common/trex_api_annotators.py", line 51, in wrap2
    ret = f(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/trex/stl/trex_stl_client.py", line 1414, in get_stats
    pgid_stats = self.get_pgid_stats()
  File "/usr/lib/python2.7/site-packages/trex/common/trex_api_annotators.py", line 51, in wrap2
    ret = f(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/trex/stl/trex_stl_client.py", line 1591, in get_pgid_stats
    return self.pgid_stats.get_stats(pgid_list)
  File "/usr/lib/python2.7/site-packages/trex/stl/trex_stl_stats.py", line 301, in get_stats
    assert rel_val >= 0, 'Negative latency error value: %s (%s %s) val=%d ref_val=%d' % (rel_val, pg_id, key, val, ref_val)
AssertionError: Negative latency error value: -2 (257 dropped) val=20481 ref_val=20483


I did recheck the nfvbench code and it does pause for 2 sec, followed by a clear_stats in between runs.


Looking at the latency returned and comparing to the hdrh:

       'latency': {'average': 44811.0,
                   'hdrh': u'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASLwvpQJAFuaBZE=',
                   'histogram': {},
                   'jitter': 5601,
                   'last_max': 0,
                   'total_max': 0,
                   'total_min': 80000}},

histogram is empty ({}) but:

(py27) AHOTHAN-M-V1YD:test ahothan$ python test_hdrh_dump.py 'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASLwvpQJAFuaBZE='
Dumping histogram: HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASLwvpQJAFuaBZE=
       Value     Percentile TotalCount 1/(1-Percentile)

   89663.000 0.000000000000          1           1.00
   89663.000 1.000000000000          1
#[Mean    =    89632.000, StdDeviation   =        0.000]
#[Max     =    89663.000, TotalCount     =        1.000]
#[Buckets =           14, SubBuckets     =         2048]

Shows 1 sample was captured with a min of 89663usec.


This one has a total_max but no last_max, also 1 sample captured (per hdrh) and the total_max seems to be incorrect:

       'latency': {'average': 54160.0,
                   'hdrh': u'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASKwvYoJAFr8BV4=',
                   'histogram': {100000: 1},
                   'jitter': 6770,
                   'last_max': 0,
                   'total_max': 156434,
                   'total_min': 100000}},

(py27) AHOTHAN-M-V1YD:test ahothan$ python test_hdrh_dump.py 'HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASKwvYoJAFr8BV4='
Dumping histogram: HISTFAAAACF4nJNpmSzMwMDAzAABMJoRTM6Y1mD/ASKwvYoJAFr8BV4=
       Value     Percentile TotalCount 1/(1-Percentile)

  108351.000 0.000000000000          1           1.00
  108351.000 1.000000000000          1
#[Mean    =   108320.000, StdDeviation   =        0.000]
#[Max     =   108351.000, TotalCount     =        1.000]
#[Buckets =           14, SubBuckets     =         2048]


This one with 3 samples looks correct:
         'latency': {'average': 286555.0,
                   'hdrh': u'HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEisL2Kaasc00ImJgCC8Qbe',
                   'histogram': {100000: 1, 400000: 2},
                   'jitter': 62897,
                   'last_max': 407681,
                   'total_max': 407681,
                   'total_min': 100000}},

(py27) AHOTHAN-M-V1YD:test ahothan$ python test_hdrh_dump.py 'HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEisL2Kaasc00ImJgCC8Qbe'
Dumping histogram: HISTFAAAACh4nJNpmSzMwMDAyQABzFCaEUzOmNZg/wEisL2Kaasc00ImJgCC8Qbe
       Value     Percentile TotalCount 1/(1-Percentile)

  108351.000 0.000000000000          1           1.00
  108351.000 0.100000000000          1           1.11
  108351.000 0.190000000000          1           1.23
  108351.000 0.271000000000          1           1.37
  407807.000 0.343900000000          2           1.52
  407807.000 0.409510000000          2           1.69
  407807.000 0.468559000000          2           1.88
  407807.000 0.521703100000          2           2.09
  407807.000 0.569532790000          2           2.32
  407807.000 0.612579511000          2           2.58
  407807.000 0.651321559900          2           2.87
  445183.000 0.686189403910          3           3.19
  445183.000 1.000000000000          3
#[Mean    =   320352.000, StdDeviation   =   150703.723]
#[Max     =   445183.000, TotalCount     =        3.000]
#[Buckets =           14, SubBuckets     =         2048]


Any suggestion on how to debug this further?

Thanks

  Alec


From: Hanoh Haim <notifications@github.com>
Reply-To: cisco-system-traffic-generator/trex-core <reply@reply.github.com>
Date: Sunday, August 25, 2019 at 5:19 AM
To: cisco-system-traffic-generator/trex-core <trex-core@noreply.github.com>
Cc: "Alec Hothan (ahothan)" <ahothan@cisco.com>, Author <author@noreply.github.com>
Subject: Re: [cisco-system-traffic-generator/trex-core] Avoid raising an exception if returned latency is -1 (#326)


Hi Alec,
I've reverted the commit,
1.  You fixed it in the wrong line. It should be in the latency test.
2.  The fix in not clear. Could you dump the rel_val and current_val to see why the value is -1?

 if 'err_cntrs' in pg_id_val and 'err_cntrs' in self.ref['latency'][pg_id]:

                    for key, val in pg_id_val['err_cntrs'].items():

                        ref_val = self.ref['latency'][pg_id]['err_cntrs'].get(key)

                        if ref_val is not None:

                            rel_val = val - ref_val

                            assert rel_val >= 0, 'Negative latency error value: %s (%s %s)' % (rel_val, pg_id, key)

                            pg_id_val['err_cntrs'][key] = rel_val

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/cisco-system-traffic-generator/trex-core/pull/326?email_source=notifications&email_token=ABFZUFNO5CAVPTY4P5WZANLQGJ2EVA5CNFSM4IPG3YCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5CSOGI#issuecomment-524625689>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABFZUFJXOX6VBEXNSD5DJILQGJ2EVANCNFSM4IPG3YCA>.

ybrustin commented:
Which version are you using?
Which NICs?
Is it bare metal?

ahothan commented:
Trex v2.59
Intel XXV710 25GbE
Trex runs in a Docker container.

This seems to happen only for slow vswitch (OVS) which drops a ton of packets when under stress. Did not see any issue so far with faster VPP-DPDK vswitch.

We never saw that issue with Intel X710.

hhaim commented:
@ahothan 
The last_max takes a few seconds to warmup so latency value would be more accurate.
So it make sense it is zero.
I don’t understand how the number of drops go backwards. Is it only this counter? (key=='err_cntrs'). If yes, we might sending diff and not absolute running number.
If this is the case it should be fixed in the server side.

ahothan commented:
It is not only dropped that decreased, seq too high also drcreased:

257: {'err_cntrs': {'dropped': 20483,        <<<<<<<<< decreased to 20481
                     'dup': 0,
                     'out_of_order': 0,
                     'seq_too_high': 1,  <<<<<<<<<<
                     'seq_too_low': 0},


257: {'err_cntrs': {'dropped': 20481,
                     'dup': 0,
                     'out_of_order': 2,
                     'seq_too_high': 0,
                     'seq_too_low': 2},

Can you explain why you had to check the counters never decreased? Did you have a case before where it happened?
Could we just pick the latency values which seem to be correct (at least in hdrh decode) and ignore the error counter drops?

hhaim commented:
@ahothan understood. The server might have issue with all the error counters.
With XL710/high speed vSwitch this situation of errors might not happening.
We are assuming that the counters are going up this is the reason for the assert.
What about the histogram/hdr? Is there negative diff there?

ahothan commented:
hdrh only records latency values, as you can see in the dumps above they all seem good.
maybe the code should just warn about decreasing counts and let the user handle the latency values returned? In all cases, the hdrh latencies are already more precise than last_max total_max and total_min and the histogram dict as well.

ahothan commented:
this exception just kills my binary search in a case where latency dont even matter...

hhaim commented:
One more question, are you using ‘’’—software’’’ in the server CLI?

ahothan commented:
I use:
./_t-rex-64 -i -c 4 --iom 0 --no-scapy-server --unbind-unused-ports --close-at-end --vlan --hdrh --mbuf-factor 0.2 --cfg /etc/trex_cfg.yaml

# cat /etc/trex_cfg.yaml
# Config generated by NFVbench
        - port_limit : 2
          version    : 2
          zmq_pub_port : 4500
          zmq_rpc_port : 4501
          prefix       : trex-local
          limit_memory : 1024
          interfaces : ['af:00.0','af:00.1']

hhaim commented:
Interesting, in this case there is only one core that handle the counters, I will look into this in the server side (it is a simple path)
I understand that for your use-case you don’t care about the errors but we should fix this for the general case.

hhaim commented:
@ahothan I've looked into the code and the guy that wrote it wanted to be accurate so some numbers could go backward **by design** for example if the right packet seq number was rx he update the errors. In short I think that for "err_cntrs" we could remove the assert as it is valid to be update back.

hhaim commented:
@ahothan @ybrustin please review the fix
